haskell/actions.bzl

```diff
@@ -633,88 +642,121 @@ def _get_library_name(ctx):
   """
-  return "HS{0}".format(get_pkg_id(ctx))
+  return "HS{0}".format(_get_pkg_id(ctx))
 
-def gather_dependency_information(ctx):
-  """Collapse dependencies into a single `HaskellPackageInfo`.
+def gather_dep_info(ctx):
+  """Collapse dependencies into a single `HaskellBuildInfo`.
 
-  `name`, `prebuilt_dependencies`, `external_libraries`, `import_dirs`,
-  `exposed_modules`, and `hidden_modules` fields are fully pre-populated.
+  Note that the field `prebuilt_dependencies` also includes
+  prebuilt_dependencies of current target.
 
   Args:
     ctx: Rule context.
 
   Returns:
-    HaskellPackageInfo: Unified information about all dependencies needed during build.
+    HaskellBuildInfo: Unified information about all dependencies.
   """
 
-  # Infer collection of public modules in the library.
-  hidden_modules = set.from_list(ctx.attr.hidden_modules) if hasattr(ctx.attr, "hidden_modules") else set.empty()
-  exposed_modules = set.empty()
-  for f in _hs_srcs(ctx):
-    mname = module_name(ctx, f)
-    if not set.is_member(hidden_modules, mname):
-      set.mutable_insert(exposed_modules, mname)
-
-  hpi = HaskellPackageInfo(
-    name = get_pkg_id(ctx),
-    names = set.empty(),
-    confs = set.empty(),
-    caches = set.empty(),
+  acc = HaskellBuildInfo(
+    package_names = set.empty(),
+    package_confs = set.empty(),
+    package_caches = set.empty(),
     static_libraries = [],
     dynamic_libraries = set.empty(),
     interface_files = set.empty(),
     prebuilt_dependencies = set.from_list(ctx.attr.prebuilt_dependencies),
     external_libraries = set.empty(),
-    import_dirs = set.singleton(import_hierarchy_root(ctx)),
-    exposed_modules = exposed_modules,
-    hidden_modules = hidden_modules,
-    haddock_ghc_args = None,
   )
 
   for dep in ctx.attr.deps:
-    if HaskellPackageInfo in dep:
-      pkg = dep[HaskellPackageInfo]
-      hpi = HaskellPackageInfo(
-        name = hpi.name,
-        names = set.mutable_insert(hpi.names, pkg.name),
-        confs = set.mutable_union(hpi.confs, pkg.confs),
-        caches = set.mutable_union(hpi.caches, pkg.caches),
-        static_libraries = hpi.static_libraries + pkg.static_libraries,
-        dynamic_libraries = set.mutable_union(hpi.dynamic_libraries, pkg.dynamic_libraries),
-        interface_files = set.mutable_union(hpi.interface_files, pkg.interface_files),
-        prebuilt_dependencies = set.mutable_union(hpi.prebuilt_dependencies, pkg.prebuilt_dependencies),
-        external_libraries = set.mutable_union(hpi.external_libraries, pkg.external_libraries),
-        import_dirs = hpi.import_dirs,
-        exposed_modules = hpi.exposed_modules,
-        hidden_modules = hpi.hidden_modules,
-        haddock_ghc_args = None,
+    if HaskellBuildInfo in dep:
+      binfo = dep[HaskellBuildInfo]
+      package_names = acc.package_names
+      if HaskellBinaryInfo in dep:
+        fail("Target {0} cannot depend on binary".format(ctx.attr.name))
```

Comment:
I actually think we should allow binaries as input somehow. For example if you need additional build tool available at build time to pre-process some files. Having said that, it has nothing to do with this PR, I was just reminded of it here. We might want to open an issue (or ignore it for now though I on a client project we have precisely such a need so current rules_haskell could not be used there).

Reply:
Let's open an issue for this.